File: js/NaturalnessModel.js
const Sequence = require('./Sequence.js').Sequence;
const NgramSuccessorModel = require('./NgramSuccessorModel.js').NgramSuccessorModel
const Ngram = require('./Ngram.js').Ngram;

const PROBA_OF_UNKNOWN = 0; //0.000001;
const DEPTH = 3;

class NaturalnessModel {
    constructor(depth, probaOfUnknown) {
        this.ngramMap = new Map();
        this.depth = depth || DEPTH;
        this.probaOfUnknown = probaOfUnknown || PROBA_OF_UNKNOWN;
    }

    crossEntropy(sequence) {
        checkSequenceType(sequence);
        if (sequence.eventList.length === 0) return this.probaOfUnknown;
        let probabilitySum = 0;
        for (let index = 0; index < sequence.eventList.length; index++) {
            let currentEvent = sequence.eventList[index];
            let currentNgram = sequence.getNgram(index, this.depth);
            let modelProba = this.getProbability(currentNgram, currentEvent);
            let proba;
            if (modelProba === 0) {
                proba = this.probaOfUnknown;
            } else {
                proba = modelProba * (1 - this.probaOfUnknown);
            }
            probabilitySum = probabilitySum + Math.log2(proba);
        }
        return -(probabilitySum / sequence.eventList.length);
    }

    getProbability(ngram, event) {
        let successor = this.ngramMap.get(ngram.key);
        if (successor == undefined) {
            return 0;
        }
        return successor.getProbability(event);
    }

    learn(sequence) {
        checkSequenceType(sequence);
        for (let index = 0; index < sequence.eventList.length; index++) {
            let ngram = sequence.getNgram(index, this.depth);
            let ngramSuccessor = this.ngramMap.get(ngram.key);
            if (ngramSuccessor === undefined) {
                ngramSuccessor = new NgramSuccessorModel();
                this.ngramMap.set(ngram.key, ngramSuccessor);           
            }
            ngramSuccessor.learn(sequence.eventList[index]);
        }
    }

    getNgramSuccessorModel(ngram) {
        checkNgramType(ngram);
        return this.ngramMap.get(ngram.key);
    }
}

function checkSequenceType(sequence) {
    if (!(sequence instanceof Sequence)) {
        throw 'sequence is not a Sequence';
    }
}

function checkNgramType(ngram) {
    if (ngram == null || ngram == undefined) {
        throw 'ngram is null or undefined';
    }
    if (!(ngram instanceof Ngram)) {
        throw 'ngram is not a Ngram';
    }
}
module.exports.NaturalnessModel = NaturalnessModel;

File: java/src/main/java/com/naturalness/NaturalnessModel.java
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

package com.naturalness;

import java.util.List;
import java.util.Map;
import java.util.HashMap;

public class NaturalnessModel {
    private final int DEPTH_DEFAULT = 3;
    private final double PROBA_OF_UNKNOWN_DEFAULT = 1e-6;

    private int depth;
    private double probaOfUnknown;
    private Map<NGram, NGramSuccessorModel> ngramMap; 

    public NaturalnessModel() {
        depth = DEPTH_DEFAULT;
        probaOfUnknown = PROBA_OF_UNKNOWN_DEFAULT;
        ngramMap = new HashMap<NGram, NGramSuccessorModel>();
    }

    public NaturalnessModel(int depth, double probaOfUnknown) {
        this();
        this.depth = depth;
        this.probaOfUnknown = probaOfUnknown;
    }

    public int getDepth() {
        return depth;
    }

    public double getProbaOfUnknown() {
        return probaOfUnknown;
    }

    public double crossEntropy(Sequence sequence) {
        if (sequence == null) {
            throw new IllegalArgumentException("sequence cannot be null");
        }
        List<Event> eventList = sequence.getEventList();
        
        if (eventList.size() == 0 ) return probaOfUnknown;

        double probabilitySum = 0;

        for (int i = 0; i < eventList.size(); i++) {
            Event currentEvent = eventList.get(i);
            NGram currentNgram = sequence.getNgram(i, depth);
            double modelProba = getProbability(currentNgram, currentEvent);
            double proba;
            if (modelProba == 0) {
                proba = probaOfUnknown;
            } else {
                proba = modelProba * (1 - probaOfUnknown);
            }
            probabilitySum = probabilitySum + Math.log(proba)/Math.log(2);
        }
        return - (probabilitySum / eventList.size());
    }

    public double getProbability(NGram ngram, Event event) {
        if (ngram == null || event == null) {
            return 0;
        }
        if (! ngramMap.containsKey(ngram)) {
            return 0;
        }
        return ngramMap.get(ngram).getProbability(event);
    }

    public void learn(Sequence sequence) {
        List<Event> eventList = sequence.getEventList();
        for (int i = 0; i < eventList.size() ; i++) {
            NGram ngram = sequence.getNgram(i, depth);
            if (!ngramMap.containsKey(ngram)) {
                ngramMap.put(ngram, new NGramSuccessorModel());
            }
            NGramSuccessorModel ngramSuccessor = ngramMap.get(ngram);
            ngramSuccessor.learn(eventList.get(i));   
        }
    }
}